cli/internal/cmd/upgradeapply.go

```diff
@@ -125,31 +123,37 @@ func (u *upgradeApplyCmd) upgradeApply(cmd *cobra.Command, fileHandler file.Hand
 			}
 		}
 	}
-	if err := handleInvalidK8sPatchVersion(cmd, conf.KubernetesVersion, flags.yes); err != nil {
+	validK8sVersion, err := validK8sVersion(cmd, conf.KubernetesVersion, flags.yes)
+	if err != nil {
 		return err
 	}
 
 	var idFile clusterid.File
-	if err := fileHandler.ReadJSON(constants.ClusterIDsFilename, &idFile); err != nil {
+	if err := u.fileHandler.ReadJSON(constants.ClusterIDsFilename, &idFile); err != nil {
 		return fmt.Errorf("reading cluster ID file: %w", err)
 	}
+	if idFile.MeasurementSalt == nil {
+		// TODO(elchead): remove after 2.10, since 2.9 does not yet save it in the idfile
+		measurementSalt, err := u.upgrader.GetMeasurementSalt(cmd.Context())
+		if err != nil {
+			return fmt.Errorf("getting join-config: %w", err)
+		}
+		idFile.MeasurementSalt = measurementSalt
```

Comment:
This is never persisted, is it?

Reply:
it is during writeOutput, but I agree that it is not so easy to see when reading the code.